### PR TITLE
Add `hasClosure` to `std/typetraits`

### PR DESCRIPTION
**Pull request by ringabout**
Unblock https://github.com/nim-lang/Nim/pull/17447

### CODE REVIEW

**Comment by Araq** [comment]
- This is a *type* trait, not an effect trait so it should be added to `typetraits`.

File: changelog.md [hunk 1]
```diff
@@ -244,6 +244,8 @@
 
 - Added dollar `$` and `len` for `jsre.RegExp`.
 
+- Added `hasClosure` to `std/effecttraits`.
```
Comment (timotheecour) [inline changelog.md]:
```suggestion
- Added `hasClosure` to `std/typetraits`.
```

File: lib/std/effecttraits.nim [hunk 2]
```diff
@@ -14,7 +14,7 @@
 ## One can test for the existance of this standard module
 ## via `defined(nimHasEffectTraitsModule)`.
 
-import macros
+import std/macros
```
Comment (Araq) [inline lib/std/effecttraits.nim]:
```suggestion
import macros
```

File: lib/pure/typetraits.nim [hunk 3]
```diff
@@ -263,3 +263,14 @@ since (1, 1):
 
     type T2 = T
     genericParamsImpl(T2)
+
+
+proc hasClosureImpl(n: NimNode): bool = discard "see compiler/vmops.nim"
+
+proc hasClosure*(fn: NimNode): bool {.since: (1, 5, 1).} =
+  ## Return true if the func/proc/etc `fn` has `closure`.
```
Comment (konsumlamm) [inline lib/pure/typetraits.nim:271]: This doc comment should be improved in the future, I for one have no idea what "if [...] `fn` has `closure`" is supposed to mean. That `fn` is a closure? Or is `closure` some field or flag?

Also, an example would be nice.

File: tests/stdlib/ttypetraits.nim [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+import std/typetraits
+
+
+
+macro testClosure(fn: typed, flag: static bool) =
+  if flag:
+    doAssert hasClosure(fn)
+  else:
+    doAssert not hasClosure(fn)
```
Comment (timotheecour) [inline tests/stdlib/ttypetraits.nim:13]: better:
```nim
macro testClosure(fn: typed): bool =
  newLit hasClosure(fn)
```
and the use it with `doAssert testClosure(fn)` etc

File: lib/pure/typetraits.nim [hunk 5]
```diff
@@ -263,3 +263,14 @@ since (1, 1):
 
     type T2 = T
     genericParamsImpl(T2)
+
+
+proc hasClosureImpl(n: NimNode): bool = discard "see compiler/vmops.nim"
+
+proc hasClosure*(fn: NimNode): bool {.since: (1, 5, 1).} =
```
Comment (timotheecour) [inline lib/pure/typetraits.nim:270]: IMO this should be `isClosure`

File: lib/pure/typetraits.nim [hunk 6]
```diff
@@ -263,3 +263,14 @@ since (1, 1):
...
+  ## Return true if the func/proc/etc `fn` has `closure`.
+  ## `fn` has to be a resolved symbol of kind `nnkSym`. This
```
Comment (timotheecour) [inline lib/pure/typetraits.nim:272]: > func/proc/etc

instead, enumerate the whole set of symbols that can be closures

File: lib/pure/typetraits.nim (hunk 5 above)
Comment (ringabout) [reply lib/pure/typetraits.nim:270]: yeah, it is related to procs not types.

**Pull request merged**